File: leetcode.com/flood-fill.py
#!/bin/python3
# https://leetcode.com/problems/flood-fill/

class Solution:
    def floodFill(self, image: list[list[int]], sr: int, sc: int, newColor: int) -> list[list[int]]:
        color = image[sr][sc]
        queue = [ (sr, sc) ]
        visited = set()

        while len(queue) > 0:
            r, c = queue.pop()

            if (r, c) in visited:
                continue
            else:
                visited.add((r, c))

            if not 0 <= r < len(image):
                continue

            if not 0 <= c < len(image[0]):
                continue

            if image[r][c] != color:
                continue

            image[r][c] = newColor

            for r_delta, c_delta in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                queue.append((r + r_delta, c + c_delta))

        return image

Solution().floodFill([[0,0,0],[0,1,1]], 1, 1, 1)

File: leetcode.com/flood-fill.rb
require 'set'

# @param {Integer[][]} image
# @param {Integer} sr
# @param {Integer} sc
# @param {Integer} color
# @return {Integer[][]}
def flood_fill(image, sr, sc, color)
    queue = [[sr, sc]]
    visited = Set.new
    ex_color = image[sr][sc]

    while queue.length > 0 do
        r, c = *queue.pop()
        next if !(r.between?(0, image.length - 1) and c.between?(0, image[0].length))
        next if visited.include?([r, c])
        next if image[r][c] != ex_color
            
        queue << [r + 1, c]
        queue << [r, c + 1]
        queue << [r - 1, c]
        queue << [r, c - 1]

        image[r][c] = color
        visited << [r, c]
    end

    return image
end

puts flood_fill([[1,1,1],[1,1,0],[1,0,1]], sr = 1, sc = 1, color = 2)